README.md

KSUID Creator
======================================================

This is a Java implementation of [Segment's K-Sortable Globally Unique Identifiers](https://github.com/segmentio/ksuid).

In summary:

*   Sorted by generation time;
*   Can be stored as a string of 27 chars;
*   Can be stored as an array of 20 bytes;
*   String format is encoded to [base-62](https://en.wikipedia.org/wiki/Base62) (0-9A-Za-z);
*   String format is URL safe and has no hyphens.

This project contains a [micro benchmark](https://github.com/f4b6a3/ksuid-creator/tree/master/benchmark) and a good amount of [unit tests](https://github.com/f4b6a3/ksuid-creator/tree/master/src/test/java/com/github/f4b6a3/ksuid).

The jar file can be downloaded directly from [maven.org](https://repo1.maven.org/maven2/com/github/f4b6a3/ksuid-creator/).

Read the [Javadocs](https://javadoc.io/doc/com.github.f4b6a3/ksuid-creator).

Also read the [KSUID release post](https://segment.com/blog/a-brief-history-of-the-uuid).

How to Use
------------------------------------------------------

Create a KSUID:

```java
Ksuid ksuid = KsuidCreator.getKsuid();
```

Create a KSUID String:

```java
String string = KsuidCreator.getKsuid().toString();
```

### Maven dependency

Add these lines to your `pom.xml`.

```xml
<!-- https://search.maven.org/artifact/com.github.f4b6a3/ksuid-creator -->
<dependency>
  <groupId>com.github.f4b6a3</groupId>
  <artifactId>ksuid-creator</artifactId>
  <version>4.1.1</version>
</dependency>
```

See more options in [maven.org](https://search.maven.org/artifact/com.github.f4b6a3/ksuid-creator).

### Modularity

Module and bundle names are the same as the root package name.

*   JPMS module name: `com.github.f4b6a3.ksuid`
*   OSGi symbolic name: `com.github.f4b6a3.ksuid`

### Segment's KSUID

The Segment's KSUID is a 160 bit long identifier (20 bytes). It consists of a 32-bit timestamp and a 128-bit randomly generated payload. Its canonical string representation is 27 characters long.

```java
// Create a KSUID
Ksuid ksuid = KsuidCreator.getKsuid();
```

Sequence of KSUIDs:

```text
24QbW6CHZpeF5q3KvFxlQpSPuR7
24QbW75lYxOmr1KZP8gi0L7WZMW
24QbW4SpzjSBTs5PMTTo5y8z793
24QbW3RXcN0OFfjTURVCNmLqJch
24QbW3BCODy9IZAjvaAPH6qgr8U
24QbW5DlcuhMlZHEXJVAOddmkeZ
24QbW6PtsLVRl4Xi9rmPAjApl1k
24QbW0TRakXKk6bykUbCSu7BJAj
24QbVznAO74F0zMaOrIShGphXdT
24QbW5HqqFlMtRqiQ1h1BHjORxf
24QbW4bWYmhDe3mZLjWcCUUZyOh
24QbW3iL26DpJmxE31QSHLan4jB
24QbW1gwBP7yClo43pJHTTb1EMU
24QbW3pKgLubTlc8xRT380eDRXb
24QbW4lHWRuUhIY0twbkTaCbQmt
24QbW0rfP04tIdUfReOcFxKaI7o

|----|--------------------|
 time       payload
```

### Sub-second KSUID

The Sub-second KSUID is a variant of Segment's KSUID. A small number of its payload bits are traded for more timestamp bits. Its main advantage is the *precision*.

The number of sub-second bits depends on `Instant.now()` [resolution](https://stackoverflow.com/questions/11452597/precision-vs-accuracy-of-system-nanotime). In JDK-8, the usual resolution is millisecond. In JDK-9+, microsecond.

Three sub-second precisions are supported: millisecond, microsecond, and nanosecond. The precision is detected at runtime.

```java
// Create a Sub-second KSUID
Ksuid ksuid = KsuidCreator.getSubsecondKsuid();
```

Sequence of Sub-second KSUIDs:

```text
24QbW5oYmzZmRWVGuv0ANLv7NxH
24QbW5ofcCGe3QtYlS60qlBmbi5
24QbW5oPZzC4nOeOtzDphnwe9ud
24QbW5omwpeVZLZNATGzcO47FcR
24QbW5ogYyeN3MIKfllQaDl04Gr
24QbW5oZDMWf3uQwk3BgOOLKyza
24QbW5ol0wQ9s0mNcvhisQipr96
24QbW5obEClijP4R7UedowRf8bo
24QbW5oD3yDaFtpZdEsmNIFAg0x
24QbW5oF304l9mj8eSdMQv5ZdsP
24QbW5o1xxkjquiqhWqT6tP4feA
24QbW5oCDHZS6AQJDmM4p22j3PH
24QbW5oI9egLkvY3iQpQXZAZTbk
24QbW5owmgSDnATUZyE4b58wSNz
24QbW5ovHJtU9kM58kDZZGGgCQ5
24QbW5oBsHnmQseJs495DpSCCze

|------|------------------|
  time       payload
```

### Monotonic KSUID

The Monotonic KSUID is another variant of Segment's KSUID. Its payload is incremented by 1 whenever the current second is equal to the previous one. Its main advantage is *speed*.

This implementation is derived from [Monotonic ULID](https://github.com/ulid/spec). It's like Segment's [`sequence.go`](https://github.com/segmentio/ksuid/blob/master/sequence.go) generator, which generates sequential KSUIDs, but there's a difference. You must pass a seed to `sequence.go` generator. In Monotonic KSUID, the seed is regenerated every second.

```java
// Create a Monotonic KSUID
Ksuid ksuid = KsuidCreator.getMonotonicKsuid();
```

Sequence of Monotonic KSUIDs:

```text
24QcJGoCufA6t80z28wBpCWdE10
24QcJGoCufA6t80z28wBpCWdE11
24QcJGoCufA6t80z28wBpCWdE12
24QcJGoCufA6t80z28wBpCWdE13
24QcJGoCufA6t80z28wBpCWdE14
24QcJGoCufA6t80z28wBpCWdE15
24QcJGoCufA6t80z28wBpCWdE16
24QcJGoCufA6t80z28wBpCWdE17
24QcJRYPLFj3bIqFnNpoP7Rv6Hs < second changed
24QcJRYPLFj3bIqFnNpoP7Rv6Ht
24QcJRYPLFj3bIqFnNpoP7Rv6Hu
24QcJRYPLFj3bIqFnNpoP7Rv6Hv
24QcJRYPLFj3bIqFnNpoP7Rv6Hw
24QcJRYPLFj3bIqFnNpoP7Rv6Hx
24QcJRYPLFj3bIqFnNpoP7Rv6Hy
24QcJRYPLFj3bIqFnNpoP7Rv6Hz
     ^ look               ^ look

|----|--------------------|
 time       payload
```

### More Examples

Create a quick KSUID:

```java
Ksuid ksuid = Ksuid.fast();
```

---

Create a KSUID from a canonical string (27 chars, base-62):

```java
Ksuid ksuid = Ksuid.from("0123456789ABCDEFGHIJKLMNOPQ");
```

---

Get the creation instant of a KSUID:

```java
Instant instant = ksuid.getInstant(); // 2014-06-05T09:06:29Z
```

```java
// static method
Instant instant = Ksuid.getInstant("0123456789ABCDEFGHIJKLMNOPQ"); // 2014-06-05T09:06:29Z
```

---

A key generator that makes substitution easy if necessary:

```java
package com.example;

import com.github.f4b6a3.ksuid.KsuidCreator;

public class KeyGenerator {
    public static String next() {
        return KsuidCreator.getKsuid().toString();
    }
}
```
```java
String key = KeyGenerator.next();
```

---

A `KsuidFactory` with `java.util.Random`:

```java
// use a `java.util.Random` instance for fast generation
KsuidFactory factory = KsuidFactory.newInstance(new Random());

// use the factory
Ksuid ksuid = factory.create();
```

---

A `KsuidFactory` with `SplittableRandom`:

```java
// use a random function that returns a long value
SplittableRandom random = new SplittableRandom();
KsuidFactory factory = KsuidFactory.newInstance(() -> random.nextLong());

// use the factory
Ksuid ksuid = factory.create();
```

---

A `KsuidFactory` with `RandomGenerator` (JDk 17+):

```java
// use a random function that returns a long value
RandomGenerator random = RandomGenerator.getDefault();
KsuidFactory factory = KsuidFactory.newInstance(() -> random.nextLong());

// use the factory
Ksuid ksuid = factory.create();
```

---

A `KsuidFactory` with `ThreadLocalRandom`:

```java
// use a random function that returns a byte array
KsuidFactory factory = KsuidFactory.newInstance((length) -> {
    final byte[] bytes = new byte[length];
    ThreadLocalRandom.current().nextBytes(bytes);
    return bytes;
});


// use the factory
Ksuid ksuid = factory.create();
```


Benchmark
------------------------------------------------------

This section shows benchmarks comparing `KsuidCreator` to `UUID.randomUUID()`.

```
---------------------------------------------------------------------------------
THROUGHPUT (operations/msec)              Mode  Cnt     Score      Error   Units
---------------------------------------------------------------------------------
UUID_randomUUID                          thrpt    5   3387.205 ±  10.224  ops/ms (1.00)
UUID_randomUUID_toString                 thrpt    5   2810.942 ±  58.598  ops/ms
-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -
Ksuid_fast                               thrpt    5  19777.514 ± 998.524  ops/ms (5.84)
Ksuid_fast_toString                      thrpt    5   2998.873 ± 192.978  ops/ms
-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -
KsuidCreator_getKsuid                    thrpt    5   2932.173 ±  59.416  ops/ms (0.87)
KsuidCreator_getKsuid_toString           thrpt    5   1626.184 ±  66.695  ops/ms
-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -
KsuidCreator_getSubsecondKsuid           thrpt    5   2963.775 ±  38.594  ops/ms (0.87)
KsuidCreator_getSubsecondKsuid_toString  thrpt    5   1550.485 ± 124.456  ops/ms
-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -
KsuidCreator_getMonotonicKsuid           thrpt    5  16123.933 ± 226.742  ops/ms (4.76)
KsuidCreator_getMonotonicKsuid_toString  thrpt    5   3020.113 ±  56.734  ops/ms
---------------------------------------------------------------------------------
Total time: 00:03:22
---------------------------------------------------------------------------------
```

System: CPU i7-8565U, 16G RAM, Ubuntu 22.04, JVM 11, rng-tools installed.

To execute the benchmark, run `./benchmark/run.sh`.

Other identifier generators
------------------------------------------------------

Check out the other ID generators from the same family:


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [Unreleased]

Nothing unreleaded.

## [4.1.1] - 2023-12-17

Adds the option to pass an Instant supplier to the public static factory methods. (by Michele Sollecito) #17

## [4.1.0] - 2022-10-22

Add a fast method to generate identifiers. #16

## [4.0.2] - 2022-09-17

Rewrite docs. #15

## [4.0.1] - 2022-08-21

Optimized comparison. #14

## [4.0.0] - 2022-07-09

Add support for RandomGenerator in Java 17. #13

## [3.1.0] - 2022-04-21

Handle clock drift. #12

## [3.0.0] - 2022-01-30

Implemented KSUID with sub-second precision. #11

## [2.3.0] - 2021-12-11

Implemented Monotonic KSUID inspired on [Monotonic ULID](https://github.com/ulid/spec).

## [2.2.0] - 2021-11-03

Optimization: the `Ksuid.compareTo()` method is at least 25% faster.

The internal field `byte[20] bytes` was replaced by two fields: `int seconds` and `byte[16] payload`.

## [2.1.2] - 2021-10-24

Regular maintenance.

## [2.1.1] - 2021-10-03

Regular maintenance.

## [2.1.0] - 2021-09-04

Add OSGi entries to Manifest.MF #8

Module and bundle names are the same as the root package name.

The OSGi symbolic name is the same as the JPMS module name: `com.github.f4b6a3.ksuid`.

## [2.0.0] - 2021-08-08

Now you call `KsuidFactory.newInstance()` and it's variants to get a new `KsuidFactory`.

### Added

-   Added `KsuidCreator.getKsuid(Instant)`
-   Added `KsuidCreator.getMsKsuid(Instant)`
-   Added `KsuidCreator.getUsKsuid(Instant)`
-   Added `KsuidCreator.getNsKsuid(Instant)`
-   Added `KsuidFactory.create(Instant)`
-   Added `KsuidFactory.newInstance()`
-   Added `KsuidFactory.newInstance(Random)`
-   Added `KsuidFactory.newInstance(Supplier<byte[]>)`
-   Added `KsuidFactory.newMsInstance()`
-   Added `KsuidFactory.newMsInstance(Random)`
-   Added `KsuidFactory.newMsInstance(Supplier<byte[]>)`
-   Added `KsuidFactory.newUsInstance()`
-   Added `KsuidFactory.newUsInstance(Random)`
-   Added `KsuidFactory.newUsInstance(Supplier<byte[]>)`
-   Added benchmark code to compare KSUID with UUID

### Removed

-   Removed `KsuidCreator.getKsuid(long)`
-   Removed `KsuidCreator.getMsKsuid(long, int)`
-   Removed `KsuidCreator.getUsKsuid(long, int)`
-   Removed `KsuidCreator.getNsKsuid(long, int)`
-   Removed `KsuidFactory.createMs()`
-   Removed `KsuidFactory.createUs()`
-   Removed `KsuidFactory.createNs()`
-   Removed `KsuidFactory.create(long)`
-   Removed `KsuidFactory.createMs(long, int)`
-   Removed `KsuidFactory.createUs(long, int)`
-   Removed `KsuidFactory.createNs(long, int)`
-   Removed `RandomGenerator` interface

## [1.1.2] - 2021-07-18

Fix for Sonatype Lift analysis report

### Fixed

-   Fixed `int` to `long` conversion in method that use nanoseconds.

## [1.1.1] - 2021-07-18

Fix for Sonatype Lift analysis report

### Fixed

-   Fixed shifts in methods that use fractions of a second (ms, us, ns).

## [1.1.0] - 2021-07-17

Simplified the use of `KsuidFactory` with other random generators.

### Added

-   Added constructors in `KsuidFactory` for random generators.

## [1.0.1] - 2021-07-17

Creates a module name be used in Java 9+.

### Added

-   Added module name for Java 9+

## [1.0.0] - 2021-06-05

Project created as an alternative Java implementation of [K-Sortable Unique IDentifier](https://github.com/segmentio/ksuid).

### Added

-   Added `Ksuid`
-   Added `KsuidCreator`
-   Added `KsuidFactory`
-   Added `RandomGenerator`
-   Added `DefaultRandomGenerator`
-   Added test cases

[unreleased]: https://github.com/f4b6a3/ksuid-creator/compare/ksuid-creator-4.1.1...HEAD
[4.1.1]: https://github.com/f4b6a3/ksuid-creator/compare/ksuid-creator-4.0.2...ksuid-creator-4.1.1
[4.1.0]: https://github.com/f4b6a3/ksuid-creator/compare/ksuid-creator-4.0.2...ksuid-creator-4.1.0
[4.0.2]: https://github.com/f4b6a3/ksuid-creator/compare/ksuid-creator-4.0.1...ksuid-creator-4.0.2
[4.0.1]: https://github.com/f4b6a3/ksuid-creator/compare/ksuid-creator-4.0.0...ksuid-creator-4.0.1
[4.0.0]: https://github.com/f4b6a3/ksuid-creator/compare/ksuid-creator-3.1.0...ksuid-creator-4.0.0
[3.1.0]: https://github.com/f4b6a3/ksuid-creator/compare/ksuid-creator-3.0.0...ksuid-creator-3.1.0
[3.0.0]: https://github.com/f4b6a3/ksuid-creator/compare/ksuid-creator-2.3.0...ksuid-creator-3.0.0
[2.3.0]: https://github.com/f4b6a3/ksuid-creator/compare/ksuid-creator-2.2.0...ksuid-creator-2.3.0

src/main/java/com/github/f4b6a3/ksuid/Ksuid.java
/*
 * MIT License
 * 
 * Copyright (c) 2021-2022 Fabio Lima
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.github.f4b6a3.ksuid;

import java.io.Serializable;
import java.nio.ByteBuffer;
import java.time.Instant;
import java.util.Arrays;
import java.util.SplittableRandom;

/**
 * A class that represents KSUIDs.
 * <p>
 * KSUID is a 160-bit value that has two parts:
 * <ul>
 * <li><b>Time</b>: a sequence of 32 bits that represent the creation minute.
 * <li><b>Payload</b>: a sequence of 128 random bits generated by a secure
 * random generator.
 * </ul>
 * <p>
 * Instances of this class are <b>immutable</b>.
 * 
 * @see <a href="https://segment.com/blog/a-brief-history-of-the-uuid/">A brief
 *      history of the UUID</a>
 */
public final class Ksuid implements Serializable, Comparable<Ksuid> {

	private static final long serialVersionUID = 3045351825700035803L;

	/**
	 * The current time.
	 */
	private final int seconds;
	/**
	 * The random bytes.
	 */
	private final byte[] payload;

	/**
	 * Number of characters of a KSUID.
	 */
	public static final int KSUID_CHARS = 27;
	/**
	 * Number of bytes of a KSUID.
	 */
	public static final int KSUID_BYTES = 20;
	/**
	 * Number of bytes of the time of a KSUID.
	 */
	public static final int TIME_BYTES = 4;
	/**
	 * Number of bytes of the payload of a KSUID.
	 */
	public static final int PAYLOAD_BYTES = 16;

	/**
	 * Number of seconds of 2014-05-13T16:53:20Z (14e8).
	 */
	public static final long EPOCH_OFFSET = 1_400_000_000L;

	static final int BASE62_RADIX = 62;
	static final char[] BASE62_ALPHABET;
	static final byte[] BASE62_MAP;

	static {

		BASE62_ALPHABET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz".toCharArray();
		BASE62_MAP = new byte[256];

		// initialize the map with -1
		Arrays.fill(BASE62_MAP, (byte) -1);

		// map the alphabets chars to values
		for (int i = 0; i < BASE62_ALPHABET.length; i++) {
			BASE62_MAP[BASE62_ALPHABET[i]] = (byte) i;
		}
	}

	static final int KSUID_INTS = KSUID_BYTES / Integer.BYTES;
	static final long INTEGER_MASK = 0x00000000ffffffffL;

	/**
	 * Creates a new KSUID.
	 * <p>
	 * Useful to make copies of KSUIDs.
	 * 
	 * @param ksuid a KSUID
	 */
	public Ksuid(Ksuid ksuid) {

		// copy the seconds
		this.seconds = ksuid.seconds;

		// copy the payload
		this.payload = new byte[PAYLOAD_BYTES];
		System.arraycopy(ksuid.payload, 0, this.payload, 0, PAYLOAD_BYTES);
	}

	/**
	 * Creates a new KSUID.
	 * <p>
	 * Seconds parameter is the number of seconds since 1970-01-01 (Unix epoch). It
	 * is internally converted to KSUID time.
	 * <p>
	 * Payload parameter must be an array of 16 bytes.
	 * 
	 * @param seconds the number of seconds since 1970-01-01
	 * @param payload an array of 16 bytes
	 * @throws IllegalArgumentException if payload is null or its length is not 16
	 */
	public Ksuid(final long seconds, final byte[] payload) {

		if (payload == null || payload.length != PAYLOAD_BYTES) {
			throw new IllegalArgumentException("Invalid payload length or null"); // null or wrong length!
		}

		// copy the seconds
		this.seconds = (int) toKsuidTime(seconds);

		// copy the payload
		this.payload = new byte[PAYLOAD_BYTES];
		System.arraycopy(payload, 0, this.payload, 0, PAYLOAD_BYTES);
	}

	Ksuid(byte[] bytes) {

		if (bytes == null || bytes.length != KSUID_BYTES) {
			throw new IllegalArgumentException("Invalid byte array length or null"); // null or wrong length!
		}

		// copy the seconds
		int secondz = 0;
		secondz |= (bytes[0x00] & 0xff) << 0x18;
		secondz |= (bytes[0x01] & 0xff) << 0x10;
		secondz |= (bytes[0x02] & 0xff) << 0x08;
		secondz |= (bytes[0x03] & 0xff) << 0x00;
		this.seconds = secondz;

		// copy the payload
		this.payload = new byte[PAYLOAD_BYTES];
		System.arraycopy(bytes, TIME_BYTES, this.payload, 0, PAYLOAD_BYTES);
	}

	Ksuid(final int[] ints) {

		if (ints == null || ints.length != KSUID_INTS) {
			throw new IllegalArgumentException("Invalid integer array length or null"); // null or wrong length!
		}

		// copy the seconds
		this.seconds = ints[0];

		// copy the payload
		this.payload = new byte[PAYLOAD_BYTES];
		for (int i = 1, j = 0; i < ints.length; i++, j += 4) {
			this.payload[j + 0] = (byte) ((ints[i] >>> 0x18) & 0xff);
			this.payload[j + 1] = (byte) ((ints[i] >>> 0x10) & 0xff);
			this.payload[j + 2] = (byte) ((ints[i] >>> 0x08) & 0xff);
			this.payload[j + 3] = (byte) ((ints[i] >>> 0x00) & 0xff);
		}
	}

	/**
	 * Returns a fast new KSUID.
	 * <p>
	 * This static method is a quick alternative to {@link KsuidCreator#getKsuid()}.
	 * <p>
	 * It employs {@link SplittableRandom} which works very well, although not
	 * cryptographically strong. It can be useful, for example, for logging.
	 * <p>
	 * Security-sensitive applications that require a cryptographically secure
	 * pseudo-random generator should use {@link KsuidCreator#getKsuid()}.
	 * 
	 * @return a KSUID
	 * @since 4.1.0
	 */
	public static Ksuid fast() {

		final SplittableRandom random = new SplittableRandom();
		final ByteBuffer buffer = ByteBuffer.allocate(Ksuid.PAYLOAD_BYTES);

		final long seconds = Instant.now().getEpochSecond();
		final byte[] payload = buffer.putLong(random.nextLong()).putLong(random.nextLong()).array();

		return new Ksuid(seconds, payload);
	}

	/**
	 * Convert a byte array into a KSUID.
	 * 
	 * @param bytes an array of 20 bytes
	 * @return a KSUID
	 * @throws IllegalArgumentException if bytes are null or its length is not 20
	 */
	public static Ksuid from(byte[] bytes) {
		return new Ksuid(bytes);
	}

	/**
	 * Convert a canonical string into a KSUID.
	 * <p>
	 * The input string must be 27 characters long and must contain only characters
	 * from base-62 alphabet.
	 * 
	 * @param string a canonical string
	 * @return a KSUID
	 * @throws IllegalArgumentException if the input string is invalid
	 * @see <a href="https://segment.com/blog/a-brief-history-of-the-uuid/">A brief
	 *      history of the UUID</a>
	 */
	public static Ksuid from(String string) {
		return fromBase62(string);
	}

	/**
	 * Convert the KSUID into a byte array.
	 * 
	 * @return a byte array.
	 */
	public byte[] toBytes() {

		byte[] bytes = new byte[KSUID_BYTES];

		// copy the seconds
		bytes[0] = (byte) ((this.seconds >>> 0x18) & 0xff);
		bytes[1] = (byte) ((this.seconds >>> 0x10) & 0xff);
		bytes[2] = (byte) ((this.seconds >>> 0x08) & 0xff);
		bytes[3] = (byte) ((this.seconds >>> 0x00) & 0xff);

		// copy the payload
		System.arraycopy(this.payload, 0, bytes, TIME_BYTES, PAYLOAD_BYTES);

		return bytes;
	}

	/**
	 * Convert the KSUID into a canonical string.
	 * <p>
	 * The output string is 27 characters long and contains only characters from
	 * base-62 alphabet.
	 * 
	 * @return a string
	 * @see <a href="https://segment.com/blog/a-brief-history-of-the-uuid/">A brief
	 *      history of the UUID</a>
	 */
	@Override
	public String toString() {
		return toBase62(this);
	}

	/**
	 * Return the instant of creation.
	 * <p>
	 * The instant of creation is extracted from the creation time.
	 * 
	 * @return the {@link Instant} of creation
	 */
	public Instant getInstant() {
		return Instant.ofEpochSecond(getTime());
	}

	/**
	 * Return the instant of creation.
	 * <p>
	 * The instant of creation is extracted from the creation time.
	 * 
	 * @param string a canonical string
	 * @return the {@link Instant} of creation
	 * @throws IllegalArgumentException if the input string is invalid
	 */
	public static Instant getInstant(String string) {
		return Instant.ofEpochSecond(getTime(string));
	}

	/**
	 * Return the creation time.
	 * <p>
	 * The number returned is equivalent to the number of seconds since 1970-01-01
	 * (Unix epoch).
	 * 
	 * @return a number of seconds
	 */
	public long getTime() {
		return toUnixTime(this.seconds);
	}

	/**
	 * Return the creation time.
	 * <p>
	 * The number returned is equivalent to the number of seconds since 1970-01-01
	 * (Unix epoch).
	 * 
	 * @param string a canonical string
	 * @return a number of seconds
	 * @throws IllegalArgumentException if the input string is invalid
	 */
	public static long getTime(String string) {
		return from(string).getTime();
	}

	/**
	 * Return the payload.
	 * <p>
	 * The payload is an array of 16 bytes (128 bits).
	 * 
	 * @return a byte array
	 */
	public byte[] getPayload() {
		final byte[] copy = new byte[PAYLOAD_BYTES];
		System.arraycopy(this.payload, 0, copy, 0, PAYLOAD_BYTES);
		return copy;
	}

	/**
	 * Return the payload.
	 * <p>
	 * The payload is an array of 16 bytes (128 bits).
	 * 
	 * @param string a canonical string
	 * @return a byte array
	 */
	public static byte[] getPayload(String string) {
		return from(string).getPayload();
	}

	/**
	 * Check if the input string is valid.
	 * <p>
	 * The input string must be 27 characters long and must contain only characters
	 * from base-62 alphabet.
	 * 
	 * @param string a canonical string
	 * @return true if valid
	 * @see <a href="https://segment.com/blog/a-brief-history-of-the-uuid/">A brief
	 *      history of the UUID</a>
	 */
	public static boolean isValid(String string) {
		return string != null && isValid(string.toCharArray());
	}

	/**
	 * Returns a hash code value for the KSUID.
	 */
	@Override
	public int hashCode() {

		final int prime = 31;
		int result = 1;

		result = prime * result + seconds;

		for (int i = 0; i < PAYLOAD_BYTES; i++) {
			result = prime * result + payload[i];
		}

		return result;
	}

	/**
	 * Checks if some other KSUID is equal to this one.
	 */
	@Override
	public boolean equals(Object other) {

		if (other == null)
			return false;
		if (other.getClass() != Ksuid.class)
			return false;

		Ksuid that = (Ksuid) other;
		if (this.seconds != that.seconds)
			return false;

		for (int i = 0; i < PAYLOAD_BYTES; i++) {
			if (this.payload[i] != that.payload[i]) {
				return false;
			}
		}

		return true;
	}

	/**
	 * Compares two KSUIDs as unsigned 160-bit integers.
	 * <p>
	 * The first of two KSUID is greater than the second if the most significant
	 * byte in which they differ is greater for the first KSUID.
	 * 
	 * @param that a KSUID to be compared with
	 * @return -1, 0 or 1 as {@code this} is less than, equal to, or greater than
	 *         {@code that}
	 */
	@Override
	public int compareTo(Ksuid that) {

		// for UNSIGNED comparison
		final int min = 0x80000000;

		final int a = this.seconds + min;
		final int b = that.seconds + min;

		if (a > b)
			return 1;
		else if (a < b)
			return -1;

		// UNSIGNED comparison of payload bytes
		for (int i = 0; i < PAYLOAD_BYTES; i++) {
			if ((this.payload[i] & 0xff) > (that.payload[i] & 0xff)) {
				return 1;
			} else if ((this.payload[i] & 0xff) < (that.payload[i] & 0xff)) {
				return -1;
			}
		}

		return 0;
	}

	/**
	 * Returns a new KSUID by incrementing the payload of the current KSUID.
	 * 
	 * When an overflow occurs in the payload 128 bits, the time is simply
	 * incremented to <b>maintain monotonicity</b>.
	 * 
	 * @return a KSUID
	 */
	public Ksuid increment() {

		// 0xff + 1 = 0x00
		final byte overflow = 0x00;

		// copy the current KSUID
		Ksuid ksuid = new Ksuid(this);

		// increment payload bytes from right to left
		for (int i = PAYLOAD_BYTES - 1; i >= 0; i--) {
			if (++ksuid.payload[i] != overflow) {
				break; // stop if did't overflow
			} else {
				if (i == 0) {
					// If the payload component overflows,
					// increment the time component by 1.
					// This event is quite rare to occur.
					final long time = ksuid.getTime() + 1;
					ksuid = new Ksuid(time, ksuid.getPayload());
				}
			}
		}

		return ksuid; // incremented copy
	}

	/**
	 * Convert the Unix time to KSUID time.
	 * <p>
	 * The 4 fist bytes of contain an unsigned time since epoch 14e8.
	 * 
	 * @param unixTime the time in seconds since 1970-01-01
	 * @return the KSUID time
	 */
	static long toKsuidTime(final long unixTime) {
		return (unixTime - EPOCH_OFFSET) & INTEGER_MASK;
	}

	/**
	 * Convert the KSUID time to Unix time.
	 * <p>
	 * The 4 fist bytes of contain an unsigned time since epoch 14e8.
	 * 
	 * @param ksuidTime the KSUID time
	 * @return the time in seconds since 1970-01-01
	 */
	static long toUnixTime(final long ksuidTime) {
		return (ksuidTime & INTEGER_MASK) + EPOCH_OFFSET;
	}

	/**
	 * Encode a KSUID into a base-62 string.
	 * 
	 * @param ksuid a KSUID
	 * @return a string
	 */
	static String toBase62(final Ksuid ksuid) {

		int[] number = ksuid.toInts();

		int b = KSUID_CHARS; // buffer index
		char[] buffer = new char[KSUID_CHARS];

		while (!isZero(number)) {
			final int[] quotient = new int[KSUID_INTS]; // division output
			final int remainder = remainder(number, BASE62_RADIX, quotient);
			buffer[--b] = BASE62_ALPHABET[remainder];
			number = quotient;
		}

		// add padding left
		while (b > 0) {
			buffer[--b] = '0';
		}

		return new String(buffer);
	}

	/**
	 * Decode a base-62 string into a KSUID.
	 * 
	 * @param string a string
	 * @return a KSUID
	 */
	static Ksuid fromBase62(final String string) {

		int[] number = new int[KSUID_INTS];
		char[] chars = toCharArray(string);

		for (int i = 0; i < chars.length; i++) {
			final int remainder = BASE62_MAP[chars[i]];
			final int[] product = multiply(number, BASE62_RADIX, remainder, true);
			number = product;
		}

		return new Ksuid(number);
	}

	static int remainder(int[] number, int divisor, int[] quotient /* division output */) {

		long temporary = 0;
		long remainder = 0;

		for (int i = 0; i < number.length; i++) {
			temporary = (remainder << 32) | (number[i] & INTEGER_MASK);
			quotient[i] = (int) (temporary / divisor);
			remainder = temporary % divisor;
		}

		return (int) remainder;
	}

	static int[] multiply(int[] number, int multiplier, int addend, boolean validate) {

		long temporary = 0;
		long overflow = addend;
		final int[] product = new int[KSUID_INTS];

		for (int i = KSUID_INTS - 1; i >= 0; i--) {
			temporary = ((number[i] & INTEGER_MASK) * multiplier) + overflow;
			product[i] = (int) temporary;
			overflow = (temporary >>> 32);
		}

		if (validate && overflow != 0) {
			throw new IllegalArgumentException("Invalid KSUID (overflow)");
		}

		return product;
	}

	int[] toInts() {
		int[] ints = new int[KSUID_INTS];

		// copy the seconds
		ints[0] = this.seconds;

		// copy the payload
		for (int i = 1, j = 0; i < ints.length; i++, j += 4) {
			ints[i] |= (this.payload[j + 0] & 0xff) << 0x18;
			ints[i] |= (this.payload[j + 1] & 0xff) << 0x10;
			ints[i] |= (this.payload[j + 2] & 0xff) << 0x08;
			ints[i] |= (this.payload[j + 3] & 0xff) << 0x00;
		}

		return ints;
	}

	private static boolean isZero(int[] number) {
		return number[0] == 0 && number[1] == 0 && number[2] == 0 && number[3] == 0 && number[4] == 0;
	}

	private static char[] toCharArray(String string) {
		if (string == null) {
			throw new IllegalArgumentException("Invalid KSUID: null");
		}
		char[] chars = string.toCharArray();
		if (!isValid(chars)) {
			throw new IllegalArgumentException(String.format("Invalid KSUID: \"%s\"", string));
		}
		return chars;
	}

	private static boolean isValid(final char[] chars) {
		if (chars == null || chars.length != KSUID_CHARS) {
			return false; // null or wrong size!
		}
		for (int i = 0; i < chars.length; i++) {
			if (BASE62_MAP[chars[i]] == -1) {
				return false; // invalid character!
			}
		}
		return true; // It seems to be OK.
	}
}
